blockchain/chain_sync/src/tipset_syncer.rs

```diff
@@ -0,0 +1,1741 @@
+                        beacon_scheduler.clone(),
+                        chainstore.clone(),
+                        bad_block_cache.clone(),
+                        full_tipset,
+                    )
+                    .await?;
+                    tracker.write().await.set_epoch(current_epoch);
+
+                    // Persist the messages in the store
+                    if let Some(m) = bundle.messages {
+                        chain::persist_objects(chainstore.blockstore(), &m.bls_msgs)?;
+                        chain::persist_objects(chainstore.blockstore(), &m.secp_msgs)?;
+                    } else {
+                        warn!("ChainExchange request for messages returned null messages");
+                    }
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+async fn validate_tipset<
+    DB: BlockStore + Send + Sync + 'static,
+    TBeacon: Beacon + Sync + Send + 'static,
+    V: ProofVerifier + Sync + Send + 'static,
+>(
+    state_manager: Arc<StateManager<DB>>,
+    beacon_scheduler: Arc<BeaconSchedule<TBeacon>>,
+    chainstore: Arc<ChainStore<DB>>,
+    bad_block_cache: Arc<BadBlockCache>,
+    full_tipset: FullTipset,
+) -> Result<(), TipsetRangeSyncerError> {
+    // TODO: Ensure that the tipset is not the genesis tipset
+
+    let epoch = full_tipset.epoch();
+    let full_tipset_key = full_tipset.key().clone();
+
+    let mut validations = FuturesUnordered::new();
+    for b in full_tipset.into_blocks() {
+        let validation_fn = task::spawn(validate_block::<_, _, V>(
+            state_manager.clone(),
+            beacon_scheduler.clone(),
+            Arc::new(b),
+        ));
+        validations.push(validation_fn);
+    }
+
+    while let Some(result) = validations.next().await {
+        match result {
+            Ok(block) => {
+                chainstore.add_to_tipset_tracker(block.header()).await;
+            }
+            Err((cid, why)) => {
+                warn!(
+                    "Validating block [CID = {}] in EPOCH = {} failed: {}",
+                    cid.clone(),
+                    epoch,
+                    why
+                );
+                match &why {
+                    TipsetRangeSyncerError::TimeTravellingBlock(_, _)
+                    | TipsetRangeSyncerError::TipsetParentNotFound => (),
+                    why => {
+                        bad_block_cache.put(cid, why.to_string()).await;
+                    }
+                }
+                return Err(why);
+            }
+        }
+    }
+    info!(
+        "Validating tipset: EPOCH = {}, KEY = {:?}",
+        epoch, full_tipset_key.cids,
+    );
+    Ok(())
+}
+
+/// Validates block semantically according to https://github.com/filecoin-project/specs/blob/6ab401c0b92efb6420c6e198ec387cf56dc86057/validation.md
+/// Returns the validated block if `Ok`.
+/// Returns the block cid (for marking bad) and `Error` if invalid (`Err`).
+async fn validate_block<
```

Comment:
Method feels quite big. Wondering if we can divvy this! But yeah am aware that there's a lot of steps. The comments in the method helps btw.

Reply:
Left this as is for now.